okhttp/src/main/java/okhttp3/internal/tls/BasicTrustRootIndex.kt

```diff
@@ -13,58 +13,41 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package okhttp3.internal.tls;
+package okhttp3.internal.tls
 
-import java.security.PublicKey;
-import java.security.cert.X509Certificate;
-import java.util.LinkedHashMap;
-import java.util.LinkedHashSet;
-import java.util.Map;
-import java.util.Set;
-import javax.security.auth.x500.X500Principal;
+import java.security.cert.X509Certificate
+import javax.security.auth.x500.X500Principal
 
-/** A simple index that of trusted root certificates that have been loaded into memory. */
-public final class BasicTrustRootIndex implements TrustRootIndex {
-  private final Map<X500Principal, Set<X509Certificate>> subjectToCaCerts;
+/** A simple index that of trusted root certificates that have been loaded into memory.  */
+class BasicTrustRootIndex(vararg caCerts: X509Certificate) : TrustRootIndex {
+  private val subjectToCaCerts = mutableMapOf<X500Principal, MutableSet<X509Certificate>>()
 
-  public BasicTrustRootIndex(X509Certificate... caCerts) {
-    subjectToCaCerts = new LinkedHashMap<>();
-    for (X509Certificate caCert : caCerts) {
-      X500Principal subject = caCert.getSubjectX500Principal();
-      Set<X509Certificate> subjectCaCerts = subjectToCaCerts.get(subject);
-      if (subjectCaCerts == null) {
-        subjectCaCerts = new LinkedHashSet<>(1);
-        subjectToCaCerts.put(subject, subjectCaCerts);
-      }
-      subjectCaCerts.add(caCert);
+  init {
+    for (caCert in caCerts) {
+      subjectToCaCerts.getOrPut(caCert.subjectX500Principal) { mutableSetOf() }.add(caCert)
     }
   }
 
-  @Override public X509Certificate findByIssuerAndSignature(X509Certificate cert) {
-    X500Principal issuer = cert.getIssuerX500Principal();
-    Set<X509Certificate> subjectCaCerts = subjectToCaCerts.get(issuer);
-    if (subjectCaCerts == null) return null;
+  override fun findByIssuerAndSignature(cert: X509Certificate): X509Certificate? {
+    val issuer = cert.issuerX500Principal
+    val subjectCaCerts = subjectToCaCerts[issuer] ?: return null
 
-    for (X509Certificate caCert : subjectCaCerts) {
-      PublicKey publicKey = caCert.getPublicKey();
+    return subjectCaCerts.firstOrNull {
       try {
-        cert.verify(publicKey);
-        return caCert;
-      } catch (Exception ignored) {
+        cert.verify(it.publicKey)
```

Comment:
it.publicKey isn't guaranteed to be cheap

```
    @Override
    public PublicKey getPublicKey() {
        /* First try to generate the key from supported OpenSSL key types. */
        try {
            OpenSSLKey pkey = new OpenSSLKey(NativeCrypto.X509_get_pubkey(mContext, this));
            return pkey.getPublicKey();
        } catch (NoSuchAlgorithmException ignored) {
        } catch (InvalidKeyException ignored) {
        }

        /* Try generating the key using other Java providers. */
        String oid = NativeCrypto.get_X509_pubkey_oid(mContext, this);
        byte[] encoded = NativeCrypto.i2d_X509_PUBKEY(mContext, this);
        try {
            KeyFactory kf = KeyFactory.getInstance(oid);
            return kf.generatePublic(new X509EncodedKeySpec(encoded));
        } catch (NoSuchAlgorithmException ignored) {
        } catch (InvalidKeySpecException ignored) {
        }

        /*
         * We couldn't find anything else, so just return a nearly-unusable
         * X.509-encoded key.
         */
        return new X509PublicKey(oid, encoded);
    }
```

Reply:
Recommendations?

Reply:
Nope, I misread, thought you had inlined inside a loop.

Reply:
Actually technically you have since it's a predicate.

Reply:
Unfortunately I think we must call `.getPublicKey()` on each object iterated over.

Reply:
Maybe one to benchmark later whether it’s worth doing this at init time and storing in a field?

Reply:
Ahhhh, like storing a `Map<X500Principal, Set<PublicKey>>` instead. Maybe. Out of scope for this PR!